pyomo/contrib/preprocessing/plugins/induced_linearity.py

```diff
@@ -0,0 +1,325 @@
+
+    """
+
+    alias('contrib.induced_linearity',
+          doc=textwrap.fill(textwrap.dedent(__doc__.strip())))
+
+    def _apply_to(self, model):
+        """Apply the transformation to the given model."""
+        _process_container(model)
+        _process_subcontainers(model)
+
+
+def _process_subcontainers(blk):
+    for disj in blk.component_data_objects(
+            Disjunct, active=True, descend_into=True):
+        _process_container(disj)
+        _process_subcontainers(disj)
+
+
+def _process_container(blk):
+    equality_tolerance = 1E-6
+    if not hasattr(blk, '_induced_linearity_info'):
+        blk._induced_linearity_info = Block()
+    eff_discr_vars = detect_effectively_discrete_vars(
+        blk, equality_tolerance)
+    # TODO will need to go through this for each disjunct, since it does
+    # not (should not) descend into Disjuncts.
+
+    # Determine the valid values for the effectively discrete variables
+    possible_var_values = determine_valid_values(blk, eff_discr_vars)
+
+    # Collect find bilinear expressions that can be reformulated using
+    # knowledge of effectively discrete variables
+    bilinear_map = _bilinear_expressions(blk)
+
+    # Relevant constraints are those with bilinear terms that involve
+    # effectively_discrete_vars
+    processed_pairs = ComponentSet()
+    for v1, var_values in possible_var_values.items():
+        v1_pairs = bilinear_map.get(v1, ())
+        for v2, bilinear_constrs in v1_pairs.items():
+            if (v1, v2) in processed_pairs:
+                continue
+            _process_bilinear_constraints(
+                blk, v1, v2, var_values, bilinear_constrs)
+            processed_pairs.add((v2, v1))
+            # processed_pairs.add((v1, v2))  # TODO is this necessary?
+
+
+def determine_valid_values(block, discr_var_to_constrs_map):
+    """Calculate valid values for each effectively discrete variable.
+
+    We need the set of possible values for the effectively discrete variable in
+    order to do the reformulations.
+
+    Right now, we select a naive approach where we look for variables in the
+    discreteness-inducing constraints. We then adjust their values and see if
+    things are stil feasible. Based on their coefficient values, we can infer a
+    set of allowable values for the effectively discrete variable.
+
+    Args:
+        block: The model or a disjunct on the model.
+
+    """
+    possible_values = ComponentMap()
+
+    for eff_discr_var, constrs in discr_var_to_constrs_map.items():
+        # get the superset of possible values by looking through the
+        # constraints
+        for constr in constrs:
+            repn = generate_standard_repn(constr.body)
+            var_coef = sum(coef for i, coef in enumerate(repn.linear_coefs)
+                           if repn.linear_vars[i] is eff_discr_var)
+            const = -(repn.constant - constr.upper) / var_coef
+            possible_vals = frozenset((const,))
```

Comment:
`frozenset` only to indicate that I don't modify the set after creating it. Otherwise, everything still works as `set`. I don't know what the convention here should be.